### ISSUE TITLE
Test failures with Python 2.7

### ISSUE THREAD

**Issue by ddebrunner** [IBMStreams/streamsx.topology#548]
I created an instance with:

```
LD_LIBRARY_PATH /home/streamsadmin/anaconda2/lib
PYTHONHOME  /home/streamsadmin/anaconda2
```

and ran:

```
cd test/java
ant -Dtopology.test.python=$HOME/anaconda2/bin/python -Dtopology.test.base.pattern='**/python/**Test.java' unittest.distributed
```

resulting in:

```
    [junit] Tests run: 3, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 596.634 sec, Thread: 1, Class: com.ibm.streamsx.topology.test.python.PublishSubscribeStringPythonTest
    [junit] Tests run: 3, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 666.853 sec, Thread: 3, Class: com.ibm.streamsx.topology.test.python.PublishSubscribeJsonPythonTest

```

My guess is the code that deals with strings/Unicode in converting tuples has not been adjusted for 2.7.

**Comment by sdickes** [comment]
After investigating this problem for a couple of days I see where the problem occurs and I have hacked around it (since I have not yet found what a real fix is for the situation). 

`#define GET_PYTHON_VALUE_AS_UTF8_AND_ASSIGN(pvalue,pattr)                \
do {                                                                     \
      Py_ssize_t size = 0;                                               \
      PyObject *utf8StringBytes  = PyUnicode_AsUTF8String(pvalue);       \
      if (utf8StringBytes == 0) { \
        SMD_flush_PyErr_Print();\
        const char* s = PyString_AsString(pvalue); \
        GET_PYTHON_VALUE_THROWIFERROR(s);                         \
        pattr.assign(s, strlen(s));                      \
      }                                                                  \
      else {                                                                  \
        const char *utf8string = PyBytes_AsString(utf8StringBytes);        \
        GET_PYTHON_VALUE_THROWIFERROR(utf8string);                         \
        pattr.assign(utf8string, strlen(utf8string));                      \
        Py_DECREF(utf8StringBytes);                                        \
      }                                                                  \
   }                                                                     \
while(0) 
`
The problem occurs when PyUnicode_AsUTF8String() is used in the above macro and returns a null pointer which is caused by this error --
TypeError: bad argument type for built-in operation

and I think this happens because the string value is not Unicode. 
The hack is to treat the value as a string if the PyUnicode_AsUTF8String() call fails. 

I do see evidence that strings are encoded in unicode but possibly during pyTupleTransform the unicode value is no longer unicode.

Any thoughts?

**Comment by ddebrunner** [comment]
When you say 'strings' what are you referring to?

**Comment by ddebrunner** [comment]
So I think this is a Python string value which in Python 2.7 is not a unicode object, however Python 2.7 supports unicode strings as well (Python 3 strings are always unicode).

We should use PyUnicode_Check to see if the value is a unicode string or not in Python 2.7

**Comment by ddebrunner** [comment]
I wonder if a cleaner approach to supporting 2.7 with the Python/C++ api would be for the code to be written to 3.7 and then have macros that redefine the 3.7 function to be calls into 2.7 code.

Would be cleaner and then automatically catch any new places where a function is invoked. Might be an issue if the Python 2.7 and 3.5 functions have the same name but different behaviour.

**Comment by ddebrunner** [comment]
Trying to fix this in pr #562

**Comment by ddebrunner** [comment]
Fixed.
